pkg/autodiscovery/configresolver/configresolver.go

```diff
@@ -6,48 +6,92 @@
 package configresolver
 
 import (
-	"bytes"
 	"context"
 	"errors"
 	"fmt"
 	"os"
+	"regexp"
 	"strconv"
+	"strings"
 
 	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
 	"github.com/DataDog/datadog-agent/pkg/autodiscovery/listeners"
 	"github.com/DataDog/datadog-agent/pkg/autodiscovery/providers/names"
 	"github.com/DataDog/datadog-agent/pkg/util/containers"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
-	"github.com/DataDog/datadog-agent/pkg/util/tmplvar"
+
+	yaml "gopkg.in/yaml.v2"
 )
 
-type variableGetter func(ctx context.Context, key []byte, svc listeners.Service) ([]byte, error)
+type variableGetter func(ctx context.Context, key string, svc listeners.Service) (string, error)
 
 var templateVariables = map[string]variableGetter{
 	"host":     getHost,
 	"pid":      getPid,
 	"port":     getPort,
 	"hostname": getHostname,
+	"env":      getEnvvar,
 	"extra":    getAdditionalTplVariables,
 	"kube":     getAdditionalTplVariables,
 }
 
+type fakeService struct{}
+
+func (f *fakeService) GetServiceID() string {
+	return ""
+}
+
+func (f *fakeService) GetTaggerEntity() string {
+	return ""
+}
+
+func (f *fakeService) GetADIdentifiers(context.Context) ([]string, error) {
+	return nil, fmt.Errorf("fakeService")
+}
+
+func (f *fakeService) GetHosts(context.Context) (map[string]string, error) {
+	return nil, fmt.Errorf("fakeService")
+}
+
+func (f *fakeService) GetPorts(context.Context) ([]listeners.ContainerPort, error) {
+	return nil, fmt.Errorf("fakeService")
+}
+
+func (f *fakeService) GetTags() ([]string, error) {
+	return nil, fmt.Errorf("fakeService")
+}
+
+func (f *fakeService) GetPid(context.Context) (int, error) {
+	return 0, fmt.Errorf("fakeService")
+}
+
+func (f *fakeService) GetHostname(context.Context) (string, error) {
+	return "", fmt.Errorf("fakeService")
+}
+
+func (f *fakeService) IsReady(context.Context) bool {
+	return false
+}
+
+func (f *fakeService) GetCheckNames(context.Context) []string {
+	return nil
+}
+
+func (f *fakeService) HasFilter(containers.FilterType) bool {
+	return false
+}
+
+func (f *fakeService) GetExtraConfig(string) (string, error) {
+	return "", fmt.Errorf("fakeService")
+}
+
 // SubstituteTemplateEnvVars replaces %%ENV_VARIABLE%% from environment
 // variables in the config init, instances, and logs config.
 // When there is an error, it continues replacing. When there are multiple
 // errors, the one returned is the one that happened first.
 func SubstituteTemplateEnvVars(config *integration.Config) error {
-	var retErr error
-
-	for _, toResolve := range dataToResolve(config) {
-		var err error
-		*toResolve, err = resolveDataWithEnvs(*toResolve)
-		if err != nil && retErr == nil {
-			retErr = err
-		}
-	}
-
-	return retErr
+	var fakeSvc fakeService
+	return substituteTemplateVariables(context.Background(), config, &fakeSvc, nil)
```

Comment:
I'm confused why a fake is used in non-test code.  Some comments to sort that out would be helpful.

In general, I'm not sure what "Refactor the config resolver" means -- it looks like it might be changing the package's behavior?  Is that the intent?  Some more detail in the commit message would be useful.

Reply:
I’ve got rid of this `fakeService` and replaced its use by the use of `nil` instead.